### PR TITLE
add pets environments and reward functions

### PR DESCRIPTION
**Pull request by runjerry**

### CODE REVIEW

File: alf/examples/mbrl_reward_functions.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+import gin
+import torch
+
+from alf.utils import common
+# implement the respective reward functions for desired environments here
+
+
+@gin.configurable
+def reward_function_for_pendulum(obs, action):
```
Comment (Haichao-Zhang) [inline alf/examples/mbrl_reward_functions.py:23]: There is already a reward function for pendulum? It seems you are trying to organizing all mbrl reward functions in a single file here.

File: alf/examples/mbrl_reward_functions.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        return cost
+
+    cost = _observation_cost(obs) + _action_cost(action)
+    reward = -cost
+    return reward
+
+
+@gin.configurable
+def reward_function_for_halfcheetah(obs, action):
+    """Function for computing reward for gym CartPole environment. It takes
```
Comment (Haichao-Zhang) [inline alf/examples/mbrl_reward_functions.py:80]: CartPole -> halfcheetah

File: alf/examples/mbrl_reward_functions.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        return cost
+
+    cost = _observation_cost(obs) + _action_cost(action)
+    reward = -cost
+    return reward
+
+
+@gin.configurable
+def reward_function_for_pusher(obs, action):
+    """Function for computing reward for gym CartPole environment. It takes
```
Comment (Haichao-Zhang) [inline alf/examples/mbrl_reward_functions.py:102]: CartPole -> Pusher

File: alf/examples/mbrl_reward_functions.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        return cost
+
+    cost = _observation_cost(obs) + _action_cost(action)
+    reward = -cost
+    return reward
+
+
+@gin.configurable
+def reward_function_for_reacher(obs, action):
+    """Function for computing reward for gym CartPole environment. It takes
```
Comment (Haichao-Zhang) [inline alf/examples/mbrl_reward_functions.py:133]: CartPole -> Reacher

File: alf/examples/mbrl_reward_functions.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        return 0.001 * torch.sum(action**2, dim=-1)
+
+    cost = _observation_cost(obs) + _action_cost(action)
+    # negative cost as reward
+    reward = -cost
+    return reward
+
+
+@gin.configurable
+def reward_function_for_cartpole(obs, action):
```
Comment (Haichao-Zhang) [inline alf/examples/mbrl_reward_functions.py:50]: About the reward functions, sometimes its association with the corresponding env/task is clear, such as pendulum, as that is a standard task from Gym.
Sometimes it might be necessary to make the association more explicit. For example, 
The cartpole reward here is not for ``CartPole-v0`` from Gym, which also a cartpole task but with discrete actions. 
Similarly for others such the halfcheetah reward etc.

File: alf/examples/mbrl_reward_functions.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+                    -1, *list(range(new_rot_perp_axis.ndim - 1))))
+            new_rot_perp_axis = tmp_rot_perp_axis.permute(
+                *list(range(1, tmp_rot_perp_axis.ndim)), 0)
+            new_rot_perp_axis /= torch.norm(
+                new_rot_perp_axis, dim=-1, keepdim=True)
+            rot_axis, rot_perp_axis, cur_end = \
+                new_rot_axis, new_rot_perp_axis, cur_end + length * new_rot_axis
+
+        cost = torch.sum(
+            torch.square(cur_end - common.get_gym_env_attr('goal')), dim=-1)
```
Comment (Haichao-Zhang) [inline alf/examples/mbrl_reward_functions.py:179]: Will this way of retrieving the goal information still correct if we have multiple parallel environment?
It seems we are using 
```gym_env = _env.envs[0].gym``` in ```get_gym_env_attr``` in this case?

File: alf/environments/gym_pets/envs/assets/half_cheetah.xml [hunk 7]
```diff
@@ -0,0 +1,95 @@
+<!-- Cheetah Model
+
+    The state space is populated with joints in the order that they are
```
Comment (Haichao-Zhang) [inline alf/environments/gym_pets/envs/assets/half_cheetah.xml:3]: For the xml files, not sure whether should check gym/mujoco license as well apart from reference to pets, if to include them.
Another possible way might be to provide pointers/scripts to download them?

**Review by Haichao-Zhang — CHANGES_REQUESTED** [review]

File: alf/environments/gym_pets/envs/pusher.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import division
+from __future__ import print_function
+from __future__ import absolute_import
```
Comment (emailweixu) [inline alf/environments/gym_pets/envs/pusher.py:17]: __future__ imports here and in other files can be removed

File: alf/examples/mbrl_reward_functions.py (hunk 6 above)
Comment (emailweixu) [reply alf/examples/mbrl_reward_functions.py:179]: same concern

File: alf/environments/gym_pets/envs/cartpole.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+from __future__ import absolute_import
+
+import os
+
+import numpy as np
+from gym import utils
+from gym.envs.mujoco import mujoco_env
+
+
+class CartpoleEnv(mujoco_env.MujocoEnv, utils.EzPickle):
```
Comment (emailweixu) [inline alf/environments/gym_pets/envs/cartpole.py:26]: need descriptions for all these new environments.
